File: src/main/java/net/pterodactylus/sone/web/ajax/GetStatusAjaxPage.java
/*
 * Sone - GetStatusAjaxPage.java - Copyright © 2010–2013 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.web.ajax;

import static com.fasterxml.jackson.databind.node.JsonNodeFactory.instance;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Collection;
import java.util.Collections;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import net.pterodactylus.sone.data.Post;
import net.pterodactylus.sone.data.PostReply;
import net.pterodactylus.sone.data.Sone;
import net.pterodactylus.sone.notify.ListNotificationFilters;
import net.pterodactylus.sone.template.SoneAccessor;
import net.pterodactylus.sone.web.WebInterface;
import net.pterodactylus.sone.web.page.FreenetRequest;
import net.pterodactylus.util.notify.Notification;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;

/**
 * The “get status” AJAX handler returns all information that is necessary to
 * update the web interface in real-time.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class GetStatusAjaxPage extends JsonPage {

	/** Date formatter. */
	private static final DateFormat dateFormat = new SimpleDateFormat("MMM d, yyyy, HH:mm:ss");

	/**
	 * Creates a new “get status” AJAX handler.
	 *
	 * @param webInterface
	 *            The Sone web interface
	 */
	public GetStatusAjaxPage(WebInterface webInterface) {
		super("getStatus.ajax", webInterface);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected JsonReturnObject createJsonObject(FreenetRequest request) {
		final Sone currentSone = getCurrentSone(request.getToadletContext(), false);
		/* load Sones. always return the status of the current Sone. */
		Set<Sone> sones = new HashSet<Sone>(Collections.singleton(getCurrentSone(request.getToadletContext(), false)));
		String loadSoneIds = request.getHttpRequest().getParam("soneIds");
		if (loadSoneIds.length() > 0) {
			String[] soneIds = loadSoneIds.split(",");
			for (String soneId : soneIds) {
				/* just add it, we skip null further down. */
				sones.add(webInterface.getCore().getSone(soneId).orNull());
			}
		}
		ArrayNode jsonSones = new ArrayNode(instance);
		for (Sone sone : sones) {
			if (sone == null) {
				continue;
			}
			jsonSones.add(createJsonSone(sone));
		}
		/* load notifications. */
		List<Notification> notifications = ListNotificationFilters.filterNotifications(webInterface.getNotifications().getNotifications(), currentSone);
		Collections.sort(notifications, Notification.CREATED_TIME_SORTER);
		/* load new posts. */
		Collection<Post> newPosts = webInterface.getNewPosts();
		if (currentSone != null) {
			newPosts = Collections2.filter(newPosts, new Predicate<Post>() {

				@Override
				public boolean apply(Post post) {
					return ListNotificationFilters.isPostVisible(currentSone, post);
				}

			});
		}
		ArrayNode jsonPosts = new ArrayNode(instance);
		for (Post post : newPosts) {
			ObjectNode jsonPost = new ObjectNode(instance);
			jsonPost.put("id", post.getId());
			jsonPost.put("sone", post.getSone().getId());
			jsonPost.put("recipient", post.getRecipientId().orNull());
			jsonPost.put("time", post.getTime());
			jsonPosts.add(jsonPost);
		}
		/* load new replies. */
		Collection<PostReply> newReplies = webInterface.getNewReplies();
		if (currentSone != null) {
			newReplies = Collections2.filter(newReplies, new Predicate<PostReply>() {

				@Override
				public boolean apply(PostReply reply) {
					return ListNotificationFilters.isReplyVisible(currentSone, reply);
				}

			});
		}
		/* remove replies to unknown posts. */
		newReplies = Collections2.filter(newReplies, PostReply.HAS_POST_FILTER);
		ArrayNode jsonReplies = new ArrayNode(instance);
		for (PostReply reply : newReplies) {
			ObjectNode jsonReply = new ObjectNode(instance);
			jsonReply.put("id", reply.getId());
			jsonReply.put("sone", reply.getSone().getId());
			jsonReply.put("post", reply.getPostId());
			jsonReply.put("postSone", reply.getPost().get().getSone().getId());
			jsonReplies.add(jsonReply);
		}
		return createSuccessJsonObject().put("loggedIn", currentSone != null).put("options", createJsonOptions(currentSone)).put("sones", jsonSones).put("notificationHash", notifications.hashCode()).put("newPosts", jsonPosts).put("newReplies", jsonReplies);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected boolean needsFormPassword() {
		return false;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected boolean requiresLogin() {
		return false;
	}

	//
	// PRIVATE METHODS
	//

	/**
	 * Creates a JSON object from the given Sone.
	 *
	 * @param sone
	 *            The Sone to convert to a JSON object
	 * @return The JSON representation of the given Sone
	 */
	private JsonNode createJsonSone(Sone sone) {
		ObjectNode jsonSone = new ObjectNode(instance);
		jsonSone.put("id", sone.getId());
		jsonSone.put("name", SoneAccessor.getNiceName(sone));
		jsonSone.put("local", sone.getInsertUri() != null);
		jsonSone.put("status", sone.getStatus().name());
		jsonSone.put("modified", webInterface.getCore().isModifiedSone(sone));
		jsonSone.put("locked", webInterface.getCore().isLocked(sone));
		jsonSone.put("lastUpdatedUnknown", sone.getTime() == 0);
		synchronized (dateFormat) {
			jsonSone.put("lastUpdated", dateFormat.format(new Date(sone.getTime())));
		}
		jsonSone.put("lastUpdatedText", GetTimesAjaxPage.getTime(webInterface, sone.getTime()).getText());
		return jsonSone;
	}

	/**
	 * Creates a JSON object that contains all options that are currently in
	 * effect for the given Sone (or overall, if the given Sone is {@code null}
	 * ).
	 *
	 * @param currentSone
	 *            The current Sone (may be {@code null})
	 * @return The current options
	 */
	private static JsonNode createJsonOptions(Sone currentSone) {
		ObjectNode options = new ObjectNode(instance);
		if (currentSone != null) {
			options.put("ShowNotification/NewSones", currentSone.getOptions().getBooleanOption("ShowNotification/NewSones").get());
			options.put("ShowNotification/NewPosts", currentSone.getOptions().getBooleanOption("ShowNotification/NewPosts").get());
			options.put("ShowNotification/NewReplies", currentSone.getOptions().getBooleanOption("ShowNotification/NewReplies").get());
		}
		return options;
	}

}

File: src/main/kotlin/net/pterodactylus/sone/web/ajax/GetStatusAjaxPage.kt
package net.pterodactylus.sone.web.ajax

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import net.pterodactylus.sone.core.ElementLoader
import net.pterodactylus.sone.core.LinkedElement
import net.pterodactylus.sone.data.Post
import net.pterodactylus.sone.data.PostReply
import net.pterodactylus.sone.data.Sone
import net.pterodactylus.sone.data.SoneOptions
import net.pterodactylus.sone.freenet.L10nFilter
import net.pterodactylus.sone.template.SoneAccessor
import net.pterodactylus.sone.text.TimeTextConverter
import net.pterodactylus.sone.utils.jsonObject
import net.pterodactylus.sone.utils.toArray
import net.pterodactylus.sone.web.WebInterface
import net.pterodactylus.sone.web.page.*
import java.text.SimpleDateFormat
import java.util.TimeZone
import javax.inject.Inject

/**
 * The “get status” AJAX handler returns all information that is necessary to
 * update the web interface in real-time.
 */
@ToadletPath("getStatus.ajax")
class GetStatusAjaxPage(webInterface: WebInterface, private val elementLoader: ElementLoader, private val timeTextConverter: TimeTextConverter, private val l10nFilter: L10nFilter, timeZone: TimeZone):
		JsonPage(webInterface) {

	@Inject constructor(webInterface: WebInterface, elementLoader: ElementLoader, timeTextConverter: TimeTextConverter, l10nFilter: L10nFilter):
			this(webInterface, elementLoader, timeTextConverter, l10nFilter, TimeZone.getDefault())

	private val dateFormatter = SimpleDateFormat("MMM d, yyyy, HH:mm:ss").apply {
		this.timeZone = timeZone
	}

	override fun createJsonObject(request: FreenetRequest) =
			getCurrentSone(request.toadletContext, false).let { currentSone ->
				createSuccessJsonObject().apply {
					this["loggedIn"] = currentSone != null
					this["options"] = currentSone?.options?.toJsonOptions() ?: jsonObject {}
					this["notificationHash"] = webInterface.getNotifications(currentSone).sortedBy { it.createdTime }.hashCode()
					this["sones"] = request.httpRequest.getParam("soneIds").split(',').mapNotNull(core::getSone).plus(currentSone).filterNotNull().toJsonSones()
					this["newPosts"] = webInterface.getNewPosts(currentSone).toJsonPosts()
					this["newReplies"] = webInterface.getNewReplies(currentSone).toJsonReplies()
					this["linkedElements"] = request.httpRequest.getParam("elements", "[]").asJson().map(JsonNode::asText).map(elementLoader::loadElement).toJsonElements()
				}
			}

	private operator fun JsonReturnObject.set(key: String, value: JsonNode) = put(key, value)
	private operator fun JsonReturnObject.set(key: String, value: Int) = put(key, value)
	private operator fun JsonReturnObject.set(key: String, value: Boolean) = put(key, value)

	private fun String.asJson() = ObjectMapper().readTree(this).asIterable()

	override val needsFormPassword = false
	override val requiresLogin = false

	private fun SoneOptions.toJsonOptions() = jsonObject {
		put("ShowNotification/NewSones", isShowNewSoneNotifications)
		put("ShowNotification/NewPosts", isShowNewPostNotifications)
		put("ShowNotification/NewReplies", isShowNewReplyNotifications)
	}

	private fun Iterable<Sone>.toJsonSones() = map { sone ->
		jsonObject {
			put("id", sone.id)
			put("name", SoneAccessor.getNiceName(sone))
			put("local", sone.isLocal)
			put("status", sone.status.name)
			put("modified", core.isModifiedSone(sone))
			put("locked", core.isLocked(sone))
			put("lastUpdatedUnknown", sone.time == 0L)
			synchronized(dateFormatter) {
				put("lastUpdated", dateFormatter.format(sone.time))
			}
			put("lastUpdatedText", timeTextConverter.getTimeText(sone.time).l10nText.let { l10nFilter.format(null, it, emptyMap()) })
		}
	}.toArray()

	private fun Iterable<Post>.toJsonPosts() = map { post ->
		jsonObject {
			put("id", post.id)
			put("sone", post.sone.id)
			put("time", post.time)
			put("recipient", post.recipientId.orNull())
		}
	}.toArray()

	private fun Iterable<PostReply>.toJsonReplies() = map { reply ->
		jsonObject {
			put("id", reply.id)
			put("sone", reply.sone.id)
			put("post", reply.postId)
			put("postSone", reply.post.get().sone.id)
		}
	}.toArray()

	private fun Iterable<LinkedElement>.toJsonElements() = map { (link, failed, loading) ->
		jsonObject {
			put("link", link)
			put("loading", loading)
			put("failed", failed)
		}
	}.toArray()

}
